### PR TITLE
[Fixes #249] Use parse5 for standard-compliant parsing

### LINKED ISSUE #249: Consider using parse5 for fully standard-compliant HTML parsing
As https://github.com/inikulin/parse5/issues/30 has now been fixed, we should consider using parse5 with cheerio, as parse5 is fully standard-compliant, whereas htmlparser2 has some known issues that make it non-compliant.

### PR DESCRIPTION
**Pull request by kkirsche**
[Fixes #249] Use parse5 for standard-compliant parsing

### CODE REVIEW

File: test/bootlint_test.js [hunk 1]
```diff
@@ -59,7 +59,7 @@ exports.bootlint = {
             [],
             'should not complain when the normal simple HTML5 doctype is used.');
         test.deepEqual(lintHtml(utf8Fixture('doctype/html5-legacy.html')),
-            [],
+            ['Document declares a non-HTML5 DOCTYPE'],
```
Comment (cvrebert) [inline test/bootlint_test.js]: Shouldn't need to change this. Either E001's implementation needs tweaking or presumably parse5 has a minor bug.

Comment (kkirsche) [inline test/bootlint_test.js]: Then parae5 has a bug.

**Comment by cvrebert** [comment]
So, the reason I hesitated to make a PR myself (was hoping to have some discussion) is that I have some concerns about switching to Whacko. Mind you, I'm not sure where currently I stand, but I think the decision merits discussion either way.

Pros:
* Completely spec-compliant HTML parsing. Cheerio's htmlparser2 has a few rare edge cases where it's not compliant/correct; see its issue tracker.

Cons:
* Whacko seems to be maintained by a single guy
* Cheerio is popular and has some level of an ecosystem around it. By contrast, there only seems to be one popular third-party package currently depending on Whacko (namely, [vulcanize](https://www.npmjs.com/package/vulcanize)), so by comparison it has virtually no ecosystem.
* It's unclear how up-to-date Whacko is with respect to upstream Cheerio ("This branch is 36 commits ahead, 294 commits behind cheeriojs:master")
* htmlparser2's edge cases haven't really impacted us so far.

Mitigating factors:
* The only maintenance Whacko really needs is merging in upstream Cheerio changes and keeping up with any parse5 updates. The former are fairly infrequent and the latter shouldn't be an issue as it's by the same author.
* It's not like we're currently hitting any gaps that necessitate frequent Cheerio updates anyway.
* Whacko is used in the TestCafé commercial product, which provides an incentive to keep it maintained.
* Whacko *very* highly compatible with mainline Cheerio, so if it became necessary to switch back to Cheerio later, it shouldn't be that painful to do so.

CC: @hnrch02, as it's ultimately his decision to make as maintainer.

**Comment by kkirsche** [comment]
Thanks. Honestly for me this is an opportunity to try to learn more than anything else so I really appreciate your explanation about the pros and cons.

**Comment by kkirsche** [comment]
X-Ref: inikulin/parse5#45

**Comment by cvrebert** [comment]
Hah, I filed https://github.com/inikulin/whacko/issues/13

**Comment by kkirsche** [comment]
haha oops. Sorry about that! Didn't mean to create any overlay, just wanted to make sure it was reported :smiley: Want me to close mine on parse5?

**Comment by hnrch02** [comment]
I'd rather stay with Cheerio.

**Pull request closed without merging**